Add observabilityStrategy option of none

leifmadsen opened the pull request:
Add observabilityStrategy of none which will result in no observability
components being deployed. Previously this option was configured as
backends.deploymentStrategy which was confusing. This change moves
backends.deploymentStrategy to a top level object, and renames it to
observabilityStrategy.

Closes: STF-458

leifmadsen commented:
test

leifmadsen commented:
recheck

leifmadsen commented:
recheck

leifmadsen commented:
recheck

leifmadsen commented:
recheck

csibbitt commented on roles/servicetelemetry/tasks/main.yml:
```diff
@@ -34,8 +34,8 @@
-    - servicetelemetry_vars.backends.deployment_strategy == 'use_community' or servicetelemetry_vars.backends.deployment_strategy == 'use_community_sub'
-    - servicetelemetry_vars.backends.deployment_strategy != 'none'
+    - observability_strategy == 'use_community' or observability_strategy == 'use_community_sub'
+    - observability_strategy != 'none'
```
` observability_strategy != 'none'`  seems redundant due to the line above it

csibbitt commented on build/stf-run-ci/tasks/deploy_stf.yml:
```diff
@@ -51,7 +52,23 @@
+        observabilityStrategy: none
+  when:
+    - service_telemetry_manifest is not defined
+    - __service_telemetry_observability_strategy == "use_community"
```
Should say "none"

leifmadsen commented:
recheck

leifmadsen commented:
Running a local deployment with the `observabilityStrategy: none` configuration using CI scripts. Running a local `recheck` here to make sure the default configuration of `use_community` is still valid. Local testing results will be provided once operational.

leifmadsen commented on build/stf-run-ci/tasks/setup_base.yml:
```diff
@@ -40,6 +40,8 @@
         source: redhat-operators
         sourceNamespace: openshift-marketplace
+  when:
+    - __service_telemetry_observability_strategy == "use_community"
```
I think this is wrong. The AMQ Interconnect Operator uses AMQ Cert Manager to create the certificates required for the routes that OpenStack connects to. I likely need to just remove this line, or add an `or .... == "none"`. In the future we may add `use_community_sub` and `use_openshift` parameters. I don't want to fully drop the `when` clause here just in case we forget to adjust it.

csibbitt replied:
Oh, yes, I agree. Good catch!

leifmadsen commented on deploy/crds/infra.watch_servicetelemetrys_crd.yaml:
```diff
@@ -88,14 +94,6 @@ spec:
-                    type: string
-                    enum:
-                    - use_community
-                    - use_community_sub
-                    - use_openshift
```
Future development can provide additional functionality.

leifmadsen commented:
I think at https://github.com/infrawatch/service-telemetry-operator/pull/294/files#diff-96c9982333293ca7ab47961dd9301889dc7e1f3d5d74c1f4399fadd0a1d2ea41L17 I should actually check for availability of the `certmanager.k8s.io` API as well.

pleimer commented:
Looks like CI can't find the correct commit again

pleimer commented:
recheck

leifmadsen commented:
More changes incoming to address the `validate_deployment.sh` not passing when observability strategy is set to `none`. Local testing for latest changes as well. Couple more commits inbound, but I think this is getting close.

leifmadsen commented:
```
$ ansible-playbook -e __service_telemetry_observability_strategy="none" build/run-ci.yaml

[...]

TASK [stf-run-ci : Deploy an instance of STF] ************************************************************************************************************************************************************************************************
included: /home/leif/src/github.com/infrawatch/service-telemetry-operator/build/stf-run-ci/tasks/deploy_stf.yml for localhost

TASK [stf-run-ci : Create default ServiceTelemetry manifest with observabilityStrategy use_community] ****************************************************************************************************************************************
skipping: [localhost]

TASK [stf-run-ci : Create default ServiceTelemetry manifest with observabilityStrategy none] *************************************************************************************************************************************************
ok: [localhost]

TASK [stf-run-ci : Show ServiceTelemetry manifest] *******************************************************************************************************************************************************************************************
ok: [localhost] => {
    "service_telemetry_manifest | from_yaml": {
        "apiVersion": "infra.watch/v1beta1",
        "kind": "ServiceTelemetry",
        "metadata": {
            "name": "default",
            "namespace": "service-telemetry"
        },
        "spec": {
            "observabilityStrategy": "none"
        }
    }
}

TASK [stf-run-ci : Create ServiceTelemetry instance] *****************************************************************************************************************************************************************************************
changed: [localhost]

TASK [stf-run-ci : Validate system is operational] *******************************************************************************************************************************************************************************************
changed: [localhost]

TASK [stf-run-ci : debug] ********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "validate_deployment.stdout_lines": [
        "Already on project \"service-telemetry\" on server \"https://api.stf.cloudops.psi.redhat.com:6443\".",
        "",
        "* [info] Waiting for QDR deployment to complete",
        "",
        "Waiting for deployment \"default-interconnect\" rollout to finish: 0 of 1 updated replicas are available...",
        "deployment \"default-interconnect\" successfully rolled out",
        "",
        "* [info] Waiting for smart-gateway deployment to complete",
        "",
        "Waiting for deployment \"default-cloud1-coll-meter-smartgateway\" rollout to finish: 0 of 1 updated replicas are available...",
        "deployment \"default-cloud1-coll-meter-smartgateway\" successfully rolled out",
        "Waiting for deployment \"default-cloud1-ceil-meter-smartgateway\" rollout to finish: 0 of 1 updated replicas are available...",
        "deployment \"default-cloud1-ceil-meter-smartgateway\" successfully rolled out",
        "",
        "* [info] Waiting for all pods to show Ready/Complete",
        "",
        "",
        "* [info] CI Build complete. You can now run tests."
    ]
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=142  changed=60   unreachable=0    failed=0    skipped=10   rescued=0    ignored=0   
```

leifmadsen commented:
recheck

leifmadsen commented:
recheck

leifmadsen commented:
I think this is ready for final review...

pleimer reviewed (approved):
I have not tested this but it looks good to me

leifmadsen commented:
Did another walk of the code and am adding a couple more controls to the stf-run-ci scripts to make things a bit cleaner on the `none` strategy.

leifmadsen commented:
Better controls were added for observabilityStrategy none.

```bash
TASK [stf-run-ci : Create default ServiceTelemetry manifest with observabilityStrategy none] *************************************************************************************************************************************************
ok: [localhost]

TASK [stf-run-ci : Show ServiceTelemetry manifest] *******************************************************************************************************************************************************************************************
ok: [localhost] => {
    "service_telemetry_manifest | from_yaml": {
        "apiVersion": "infra.watch/v1beta1",
        "kind": "ServiceTelemetry",
        "metadata": {
            "name": "default",
            "namespace": "service-telemetry"
        },
        "spec": {
            "observabilityStrategy": "none"
        }
    }
}

TASK [stf-run-ci : Create ServiceTelemetry instance] *****************************************************************************************************************************************************************************************
changed: [localhost]

TASK [stf-run-ci : Validate system is operational] *******************************************************************************************************************************************************************************************
changed: [localhost]

TASK [stf-run-ci : debug] ********************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "validate_deployment.stdout_lines": [
        "Already on project \"service-telemetry\" on server \"https://api.stf.cloudops.psi.redhat.com:6443\".",
        "",
        "* [info] Waiting for QDR deployment to complete",
        "",
        "Waiting for deployment \"default-interconnect\" rollout to finish: 0 of 1 updated replicas are available...",
        "deployment \"default-interconnect\" successfully rolled out",
        "",
        "* [info] Waiting for smart-gateway deployment to complete",
        "",
        "Waiting for deployment \"default-cloud1-coll-meter-smartgateway\" rollout to finish: 0 of 1 updated replicas are available...",
        "deployment \"default-cloud1-coll-meter-smartgateway\" successfully rolled out",
        "Waiting for deployment \"default-cloud1-ceil-meter-smartgateway\" rollout to finish: 0 of 1 updated replicas are available...",
        "deployment \"default-cloud1-ceil-meter-smartgateway\" successfully rolled out",
        "",
        "* [info] Waiting for all pods to show Ready/Complete",
        "",
        "",
        "* [info] CI Build complete. You can now run tests."
    ]
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=104  changed=35   unreachable=0    failed=0    skipped=26   rescued=0    ignored=0   
```

Resulting deployment.

```bash
NAME                                                      READY   STATUS      RESTARTS   AGE
default-cloud1-ceil-meter-smartgateway-68d99ccd4d-v4wz5   2/2     Running     0          5m44s
default-cloud1-coll-meter-smartgateway-64fb64db9b-n9fr2   2/2     Running     0          6m42s
default-cloud1-sens-meter-smartgateway-77dc77b9d9-8s5cw   2/2     Running     0          4m43s
default-interconnect-5657954949-7vnj7                     1/1     Running     0          7m34s
interconnect-operator-b8f5bb647-nj4fx                     1/1     Running     0          41m
prometheus-webhook-snmp-2-build                           0/1     Completed   0          11m
service-telemetry-operator-2-build                        0/1     Completed   0          41m
service-telemetry-operator-7d9b9cf6cd-ftxtt               1/1     Running     0          9m
sg-bridge-2-build                                         0/1     Completed   0          16m
sg-core-2-build                                           0/1     Completed   0          32m
smart-gateway-operator-2-build                            0/1     Completed   0          36m
smart-gateway-operator-f49ffc48d-5fkg2                    1/1     Running     0          9m9s
```

leifmadsen commented:
recheck

csibbitt commented on roles/servicetelemetry/tasks/main.yml:
```diff
@@ -12,9 +12,15 @@
-  when: has_interconnect_api | bool
+  when:
+    - has_interconnect_api | bool
+    - has_certmanager_api | bool
```
This might help with a minor annoyance I found in dev when forgetting to deploy the certmanager

leifmadsen replied:
I wonder if adding some sort of debugging information with a sort of "error" style warning would be useful here in the logs? I could add an `ignore_errors: true` on a log output with a warning style level.

I won't do it for this PR (it's already about 4 times larger than I expected), but I think it might be worth having some of those checks and missing interfaces called out a bit more strongly in the console output.

csibbitt commented:
This looks great, thanks! A bigger diff than I'd have imagined, bravo for it still being readable.

PS. Looking forward to `observabilityStrategy: use_openshift`! :grinning:

csibbitt reviewed (approved):
LGTM, testing output is pretty convincing. Go for it!
